### DISCUSSION TITLE
Change Materials On Individual Parts Using Variants

### DISCUSSION THREAD

**Discussion by WadeMorrison** [google/model-viewer#2838] [Q&A] [answered]
Hello! 

I have a basic configurator that I'm working on to display furniture products that my company offers. Right now I am using the "Variants Extension" and it works great! 

Here is a WIP of what I have so far: [Rios Chest](https://rioschest.s3.us-east-2.amazonaws.com/index.html)

However, if I wanted to have a secondary option to change just the "Drawer Face" material, how would I go about doing that? Since right now, the variant extension changes everything as a whole (the specified parts anyways). I really like what [R2U](https://github.com/google/model-viewer/discussions/1820?sort=new) showcased using model-viewer, and am trying to replicate something like that where they are able to change the fabric and wood individually.

**Comment by ghost** [comment]
Simply put the model needs to be authored so that the drawer face has its own material, then you can have a variant for that material with all the different selections. The model in the example you cite has at least 2 materials with their own set of variants.

**Reply by WadeMorrison** [reply]
@timmmeh Ah, okay that makes sense then! I'll try that out and see how it goes, thank you for the help!

**Comment by WadeMorrison** [comment]
@timmmeh Hello!

So I went ahead and created another product that now has two separate materials. I am able to adjust each one individually as well! However, whenever I go to click on a different "Wood" material, the model will revert the "Lacquer" color to it's default color. And vice versa. It also seems like the color will not change now unless I select it first, then click on the model-viewer "window". Where as previously once I clicked on the material it would automatically apply that material.

The model isn't optimized fully as I'm still trying to make a good template for the other products we sell. So loading time take's a little while, sorry! But here is another WIP of what I have so far if you'd like to dig around and see what I may be doing wrong: [Cash Credenza](https://cashcredenza.s3.us-east-2.amazonaws.com/index.html)

Again, thank you for the help you've already offered, it is much appreciated as I'm still relatively new to programming!

**Reply by ghost** [reply]
Hi, it looks like you're getting closer. I was noticing some strange errors when switching variants, some null materials were coming through and causing problems. I converted the file from glb to gltf (to inspect it) and I see that there are 4 materials defined in the file but the variant definitions for the various mesh parts show they trying to address materials indices all the way up to 25. I think the file may have be authored to have 26 materials total but they are not getting exported, so this looks like a different problem on the exporter side, essentially all 26 of those materials need to be defined in the glTF file.

**Reply by ghost** [reply]
Actually, never-mind, it looks like the `gltf-pipeline` tool removed all but 4 of the materials when converting from glb to gltf, not sure why it's doing that. When I bring the model in Model-Viewer (experimental variant version I have) I can pick each variant and everything works fine.

**Reply by ghost** [reply]
Ok so I can't actually repro the issue you are seeing it's actually working for me in our version of the Model-Viewer editor, we have made some changes to variants recently and those changes have not shipped. Are you building against the public Model-Viewer release or against Model-Viewer repo head?

**Reply by WadeMorrison** [reply]
Hello! Thank you for the quick reply!

I was going to say I am confused on what you mean by not having all materials defined in the GLTF as I though I had done that. But maybe I am possibly doing that wrong? I tried following the same-ish flow that the other variant example had (sheen chair) When I use the variant script that is up on the example model-viewer page it works great! I just was trying to find a way to where I could have it not auto-populate/auto-create the buttons for me that way I can arrange the buttons how I wanted in my HTML document. So maybe I made the variant script wrong?

**Reply by WadeMorrison** [reply]
I have no idea what that means haha, I am so sorry. I'm just using the copy/paste HTML code from the model-viewer website. And then editing it from there to fit my needs. Slowly but surely anyways. If wanted, I could share my files with you so you can dig around? Not sure if that would help or not.

**Reply by ghost** [reply]
The good news is the file looks like it's correct I can switch to each variant, changing color does not change the wood and vice versa, so yeah I'd say the issue is probably in the script you are using to do the switching.

**Reply by WadeMorrison** [reply]
Ah okay, that is good news then! And well, I will continue to work on it then and see what I can come up with. I just tried throwing it up on glitch and used the default variant example and it still reverts back to the default color of lacquer once I choose a different wood material and vise versa. Is it working on your side because you're using an experimental version? Or was this working for you on the current version?

**Reply by elalish** [reply]
It sounds like you  may be trying to use pre-release features. You can test the  latest by using this: `https://modelviewer.dev/node_modules/@google/model-viewer/dist/model-viewer.min.js` instead of `unpkg`. Only for testing! This  gets updated every time we merge a PR, so it's not for production. We'll  have a new release out soon  with  all our latest features.

**Reply by ghost** [reply]
Thanks @elalish, I was going to suggest that too I think this may be an actual model-viewer bug that is fixed in pre-release.

**Comment by WadeMorrison** [comment, marked as answer]
Good morning! Thank you both for the help, and very responsive! I will go ahead and try the testing version out. As I'm the only 3D artist at this company, and programming is a whole new world to me I'm glad to know I was headed at least in the right direction some what haha. 

Do you have any clues as to when another release of model-viewer will be by chance?

**Reply by elalish** [reply]
Aiming for next week; stay tuned.

**Reply by WadeMorrison** [reply]
I am happy to report that it is working great so far! The only bug that I notice currently is that if I change just the "Lacquer" color, and then choose a different size, it stays the same. But as soon as I choose a "Lacquer" color _and_ "Wood" color, it'll revert the lacquer color to it's default when I choose a different size. 

But maybe that's on my end, I am not sure haha. Either way, it works pretty good currently! Will look forward to next week then!

**Reply by ghost** [reply]
Curious about your variant authoring process? Are you using Blender? How do you generate a new variant?

**Reply by WadeMorrison** [reply]
Yes I use Blender, there is an add-on that [takahirox](https://github.com/takahirox/glTF-Blender-IO-materials-variants) had made, it doesn't export it out correctly but gets it about 80% of the way there, so I have to open it in VSCode to edit/correct it and export it as a GLB.

I am curious though, would it be possible to have an external material library that we could pull from like we can with textures, and target specific meshes? I think that could save a lot of time of having to set up the GLTF(s) for things like configurators. But I don't have the slightest clue on how to go about doing that if it is already possible.

**Reply by elalish** [reply]
MV only  supports changing  materials,  not splitting or combining them  on meshes. That gets  pretty complicated and it's better  to have a full authoring tool like Blender for those workflows. Basically you want  to author it with separate materials for each part you want changeable in MV afterward.

**Reply by WadeMorrison** [reply]
Alright, that's what I figured too. The configurator program my company uses now lets us pull in a material from a material database by an external identifier/specific name, it's all programmed through JSON to decide which material can go to which mesh, but they require .OBJ models. So I wasn't sure if MV could achieve the same thing using an external material library that I create. But assumed it would be pretty complicated for the current version of MV and I don't have great knowledge on "hacking" my way down to the three.js layer and I don't want to risk my app breaking with a new release haha. 

Thank you for the clarification!

**Reply by ghost** [reply]
Thanks for the info, very interesting!
You _can_ always change material properties and textures directly as in the [examples](https://modelviewer.dev/examples/scenegraph) no need to touch three.js, but you would still have to provide the code layer that reads from your internal database to fetch the texture and material setting and then apply those changes to a model in ModelViewer also you would need a unique material for each switchable part on the model, same as with variants.

**Reply by WadeMorrison** [reply]
You're welcome! I hope you don't mean interesting in a bad way haha, I've taught most of this myself through tutorials and picking apart various models/programs so if I am doing something wrong please do tell! I'm actually trying to convert my static page into a MERN stack currently which is kinda fun to make it more dynamic for the various models I have, but I've hit many bumps so far haha. Thank you for the help and clarifying on how to possibly do an internal database!

I will have to look more into the internal database fetch and see if I can figure it out somehow. This would just be easier for me down the road in case I have to edit a material or even add more selection(s) to the model(s). Then I won't have to open up the affected GLTF(s), edit it, then re-export them out. But, so far the current method is working great as is!

**Reply by ghost** [reply]
Haha, no no, interesting as in its good to get a sense of the state of asset-pipelines out there and what is working and not working.
It sounds like your right on track 👍
